resources/assets/lib/components/score-pin.tsx

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import ScoreJson from 'interfaces/score-json';
+import { computed, makeObservable } from 'mobx';
+import { observer } from 'mobx-react';
+import core from 'osu-core-singleton';
+import * as React from 'react';
+import { onErrorWithCallback } from 'utils/ajax';
+
+interface Props {
+  className: string;
+  onUpdate?: () => void;
+  score: ScoreJson;
+}
+
+@observer
+export default class ScorePin extends React.Component<Props> {
+  private onUpdateCallbacks: {
+    done: Props['onUpdate'];
+    fail: ReturnType<typeof onErrorWithCallback>;
+  } | null = null;
+
+  @computed
+  private get isPinned() {
+    return core.scorePins.isPinned(this.props.score);
+  }
+
+  @computed
+  private get label() {
+    const targetState = this.isPinned ? '0' : '1';
+
+    return osu.trans(`users.show.extra.top_ranks.pin.to_${targetState}`);
+  }
+
+  constructor(props: Props) {
+    super(props);
+
+    makeObservable(this);
+  }
+
+  componentWillUnmount() {
+    this.onUpdateCallbacks = null;
```

Comment:
the active menu key still needs to be `null`ed; `PopupMenuPersistent` unmount seems to be the appropriate place for this, although `useEffect` seems to trigger too many times for each change 🤔